banzai/qc/pattern_noise.py

```diff
@@ -21,49 +22,197 @@ def group_by_keywords(self):
 
     def do_stage(self, images):
         for image in images:
-            # If the data is a cube, then run on each extension individually
             logging_tags = logs.image_config_to_tags(image, self.group_by_keywords)
             logs.add_tag(logging_tags, 'filename', os.path.basename(image.filename))
             logs.add_tag(logging_tags, 'snr_threshold', self.SNR_THRESHOLD)
-            logs.add_tag(logging_tags, 'freq_bins_threshold', self.FREQ_BINS_THRESHOLD)
+            logs.add_tag(logging_tags, 'std_threshold', self.STD_THRESHOLD)
+
+            # If the data is a cube, then run on each extension individually
             if image.data_is_3d():
-                pattern_noise_is_bad = any([check_for_pattern_noise(data, self.SNR_THRESHOLD, self.FREQ_BINS_THRESHOLD)
-                                            for data in image.data])
+                pattern_noise_is_bad = any([self.check_for_pattern_noise(data) for data in image.data])
             else:
-                pattern_noise_is_bad = check_for_pattern_noise(image.data, self.SNR_THRESHOLD, self.FREQ_BINS_THRESHOLD)
+                pattern_noise_is_bad = self.check_for_pattern_noise(image.data)
+
             if pattern_noise_is_bad:
                 self.logger.error('Image found to have pattern noise.', extra=logging_tags)
             else:
                 self.logger.info('No pattern noise found.', extra=logging_tags)
             self.save_qc_results({'pattern_noise.failed': pattern_noise_is_bad,
                                   'pattern_noise.snr_threshold': self.SNR_THRESHOLD,
-                                  'pattern_noise.freq_bins_threshold': self.FREQ_BINS_THRESHOLD}, image)
+                                  'pattern_noise.std_threshold': self.STD_THRESHOLD,
+                                  }, image)
         return images
 
+    def check_for_pattern_noise(self, data):
+        """
+        Test for pattern noise in an image
+
+        Parameters
+        ----------
+        data : numpy array
+               Image data to test for pattern noise
+        band_width : float
+                The fractional width of the central band to extract from the FFT
+        snr_clip_fraction: float
+                The fraction of pixels to clip from each SNR edge
+
+        Returns
+        -------
+        is_bad : bool
+                 Returns true if the image has pattern noise
+
+        """
+
+        trimmed_data = trim_image_edges(data)
+        power_2d = get_2d_power_band(trimmed_data)
+        snr = compute_snr(power_2d)
+        convolved_snr = convolve_snr_with_wavelet(snr)
+        peak_maxima, std_maxima = get_peak_parameters(convolved_snr)
+
+        # Check that all peaks are above threshold and that the peak center standard deviation is small.
+        has_pattern_noise = (peak_maxima > self.SNR_THRESHOLD).all() and std_maxima < self.STD_THRESHOLD
```

Comment:
What happens if no peaks are found?

Reply:
Is there a case you have in mind where this might happen? `peak_maxima` and `std_maxima` come from `np.max` and `np.std` which are pretty robust. I added a test where the image is made of NaNs and even then `check_for_pattern_noise` runs all the way through without a problem.